<a href="https://colab.research.google.com/github/VOL-EM2024/mo/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 3

##Обработка пропущенных значений

Обработайте пропущенные значения:


* Удалите пропущенные значения
* Заполните при помощи медианы, среднего для числовых значений
* Заполните при помощи моды для категориальных значений
* Интерполируйте
Выведите статистку после каждого метода и объясните


In [49]:
!pip install scikit-learn

In [50]:
import pandas as pd

df = pd.read_csv('/content/global-data-on-sustainable-energy (1).csv')
df

Entity    Year  Access to electricity (% of population)  \
0     Afghanistan  2000.0                                 1.613591   
1     Afghanistan  2001.0                                 4.074574   
2     Afghanistan  2002.0                                 9.409158   
3     Afghanistan  2003.0                                14.738506   
4     Afghanistan  2004.0                                20.064968   
...           ...     ...                                      ...   
3644     Zimbabwe  2016.0                                42.561730   
3645     Zimbabwe  2017.0                                44.178635   
3646     Zimbabwe  2018.0                                45.572647   
3647     Zimbabwe  2019.0                                46.781475   
3648     Zimbabwe  2020.0                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
4                                  10.9   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                                NaN   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0

In [51]:
df_cleaned = df.dropna()
df_cleaned #Удаление

Entity    Year  Access to electricity (% of population)  \
43       Algeria  2001.0                                 98.96687   
44       Algeria  2002.0                                 98.95306   
45       Algeria  2003.0                                 98.93401   
46       Algeria  2004.0                                 98.91208   
47       Algeria  2005.0                                 98.88961   
...          ...     ...                                      ...   
3559  Uzbekistan  2015.0                                100.00000   
3560  Uzbekistan  2016.0                                100.00000   
3561  Uzbekistan  2017.0                                100.00000   
3562  Uzbekistan  2018.0                                100.00000   
3563  Uzbekistan  2019.0                                100.00000   

      Access to clean fuels for cooking  \
43                                97.30   
44                                97.80   
45                                98.00   
46                                98.20   
47                                98.50   
...                                 ...   
3559                              85.35   
3560                              85.20   
3561                              84.90   
3562                              84.30   
3563                              84.60   

      Renewable-electricity-generating-capacity-per-capita  \
43                                                 8.79      
44                                                 8.68      
45                                                 8.57      
46                                                 8.46      
47                                                 8.34      
...                                                 ...      
3559                                              60.83      
3560                                              59.88      
3561                                              58.24      
3562                                              59.09      
3563                                              57.96      

      Financial flows to developing countries (US $)  \
43                                          810000.0   
44                                          310000.0   
45                                           90000.0   
46                                          140000.0   
47                                          160000.0   
...                                              ...   
3559                                        270000.0   
3560                                        690000.0   
3561                                      60130000.0   
3562                                      84130000.0   
3563                                      65940000.0   

      Renewable energy share in the total final energy consumption (%)  \
43                                                 0.43                  
44                                                 0.51                  
45                                                 0.47                  
46                                                 0.44                  
47                                                 0.58                  
...                                                 ...                  
3559                                               1.71                  
3560                                               1.61                  
3561                                               1.75                  
3562                                               1.49                  
3563                                               1.57                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
43                                  24.96                             0.0   
44                                  25.94                             0.0   
45                                  27.54                             0.0   
46                                  29.14        

Мы удалили все строки, которые содержали  NaN. Если посмотреть на первую таблицу, можем увидить, что изначально строк было 3649, а после наших действий стало 324.Следовательно мы удалили строки).

In [52]:
df_cleaned.gdp_growth.describe()

count    324.000000
mean       4.492573
std        4.413222
min      -36.658153
25%        2.597033
50%        4.391612
75%        6.504695
max       34.500000
Name: gdp_growth, dtype: float64

In [53]:
df_median = df.fillna(df.median())
df_median #Заполнение при помощи медианы


<ipython-input-53-fb903157e4ef>:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_median = df.fillna(df.median())


Entity    Year  Access to electricity (% of population)  \
0     Afghanistan  2000.0                                 1.613591   
1     Afghanistan  2001.0                                 4.074574   
2     Afghanistan  2002.0                                 9.409158   
3     Afghanistan  2003.0                                14.738506   
4     Afghanistan  2004.0                                20.064968   
...           ...     ...                                      ...   
3644     Zimbabwe  2016.0                                42.561730   
3645     Zimbabwe  2017.0                                44.178635   
3646     Zimbabwe  2018.0                                45.572647   
3647     Zimbabwe  2019.0                                46.781475   
3648     Zimbabwe  2020.0                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
4                                  10.9   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                          5570000.0   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0

Мы заменили значение NaN при помощи медианы.

In [ ]:
df_median.gdp_growth.describe()

In [ ]:
df_mean = df.fillna(df.mean())
df_mean #Заполнение при помощи среднего

Тк нужно заменить NaN, мы используем метод заполнения средним.

In [ ]:
df_mean.gdp_growth.describe()

In [ ]:
df_mode = df.fillna(df["gdp_growth"].mode()[0])
df_mode #Заполнение при помощи моды

Мы заменили значение NaN при помощи моды.

In [ ]:
df_mode.gdp_growth.describe()

In [ ]:
df_interpolated = df.interpolate()
df_interpolated #Заполнение при помощи интерполяции

Используем интерполяцию для заполнения пропущенных значений.(Интерполяция – способ нахождения промежуточных значений, между двумя известными значениями.)

In [ ]:
df_interpolated.gdp_growth.describe()

Вывод: исходя из статистики, мы можем сделать вывод о том, что в данной ситуации не важно каким из способов обрабатывать пропущенные значения, так как начальные данные после применения этих методов почти не изменились.

##Поиск выбросов и аномалий

**Те, кто выполняет задания со * делают поиск выбросов с помощью написанного ранее класса**

Найти выбросы по числовой переменной (использовать метод основанный на среднем и на межквартильном размахе). Объясните результат

Вывести статистику с выбросами и без. Сравните и объясните результат


In [ ]:
import matplotlib.pyplot as plt
df = df.dropna()
plt.boxplot(df["gdp_growth"])
plt.show() #Методом среднего

выбросы есть

In [ ]:
df["gdp_growth"].describe()

In [ ]:
mean = df["gdp_growth"].mean()
std_deviation = df["gdp_growth"].std()
threshold = 3 * std_deviation

# Идентифицируем выбросы
outliers = [x for x in df["gdp_growth"] if abs(x - mean) > threshold]

# Выводим результаты
print("Среднее значение данных:", mean)
print("Стандартное отклонение данных:", std_deviation)
print("Потенциальные выбросы:", outliers) #Методом среднего

In [ ]:
Q1 = df["gdp_growth"].quantile(0.25)
Q3 = df["gdp_growth"].quantile(0.75)
IQR = Q3-Q1
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR

outliers2 = [x1 for x1 in df["gdp_growth"] if ((x1 < lower) or (x1 > upper))]#При помощи межквартильного размаха

print("Потенциальные выбросы:", outliers2)

Вывод: после поиска выбросов двумя разными методами,мы учидели, что и в 1 и во 2 мтоде выбросами являются 27.96153807, 34.5, -36.65815267.

**Объяснения**

##Дубликаты

Проверьте датафрейм на дубликаты и удалите их, если дублей нет - создайте.

In [ ]:
df_duplicated = df[df.duplicated()]
df_duplicated #Проверка на дубли

Дублей нет, поэтому их нужно создать.

In [ ]:
df.loc[len(df)] = df.iloc[1]
df #Создание дублей

In [ ]:
df_duplicated = df[df.duplicated()]
df_duplicated #Ещё раз проверяем

In [ ]:
df_duplicates = df.drop_duplicates()
df_duplicates #Удаление дублей

мы добавили, и мы удалили...

## Преобразование признаков

In [ ]:
# получаем числовые признаки
num_features = df.select_dtypes(include=['float64']).columns

num_features

In [ ]:
df.gdp_growth.unique()

Масштабируйте числовые признаки через MinMaxScaler, StandardScaler

In [ ]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# MinMax масштабирование
scaler_minmax = MinMaxScaler()
df_minmax_scaled = df.copy()
df_minmax_scaled[num_features] = scaler_minmax.fit_transform(df[num_features])
df_minmax_scaled.gdp_growth.unique() # преобразованные уникальные значения в числовом призанке в диапозоне [0, 1]


In [ ]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler
# Standard масштабирование
standard_scaler = StandardScaler()
df_standard_scaler = df.copy()
df_standard_scaler[num_features] = standard_scaler.fit_transform(df[num_features])
df_standard_scaler.gdp_growth.unique()

Масштабируйте категориальные признаки через OneHotEncoder

In [ ]:
df.info()

In [73]:
num_features = df.select_dtypes(include=['category']).columns

num_features

Index([], dtype='object')

In [ ]:
df["Entity"] = df["Entity"].astype("category")

cat_features = df.select_dtypes(include=['category']).columns

from sklearn.preprocessing import OneHotEncoder

# Применяем OneHotEncoder
encoder = OneHotEncoder(sparse=False, drop='first')
encoded_features = encoder.fit_transform(df[cat_features])

# Получаем имена новых признаков
encoded_columns = encoder.get_feature_names_out(cat_features)
encoded_columns

In [81]:
encoded = pd.get_dummies(df, columns=cat_features, drop_first=True)
encoded

Year  Access to electricity (% of population)  \
43    2001.0                                 98.96687   
44    2002.0                                 98.95306   
45    2003.0                                 98.93401   
46    2004.0                                 98.91208   
47    2005.0                                 98.88961   
...      ...                                      ...   
3559  2015.0                                100.00000   
3560  2016.0                                100.00000   
3561  2017.0                                100.00000   
3562  2018.0                                100.00000   
3563  2019.0                                100.00000   

      Access to clean fuels for cooking  \
43                                97.30   
44                                97.80   
45                                98.00   
46                                98.20   
47                                98.50   
...                                 ...   
3559                              85.35   
3560                              85.20   
3561                              84.90   
3562                              84.30   
3563                              84.60   

      Renewable-electricity-generating-capacity-per-capita  \
43                                                 8.79      
44                                                 8.68      
45                                                 8.57      
46                                                 8.46      
47                                                 8.34      
...                                                 ...      
3559                                              60.83      
3560                                              59.88      
3561                                              58.24      
3562                                              59.09      
3563                                              57.96      

      Financial flows to developing countries (US $)  \
43                                          810000.0   
44                                          310000.0   
45                                           90000.0   
46                                          140000.0   
47                                          160000.0   
...                                              ...   
3559                                        270000.0   
3560                                        690000.0   
3561                                      60130000.0   
3562                                      84130000.0   
3563                                      65940000.0   

      Renewable energy share in the total final energy consumption (%)  \
43                                                 0.43                  
44                                                 0.51                  
45                                                 0.47                  
46                                                 0.44                  
47                                                 0.58                  
...                                                 ...                  
3559                                               1.71                  
3560                                               1.61                  
3561                                               1.75                  
3562                                               1.49                  
3563                                               1.57                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
43                                  24.96                             0.0   
44                                  25.94                             0.0   
45                                  27.54                             0.0   
46                                  29.14                             0.0   
47                                  31.36                             0.0   
...                                  

Ивините, но это ужас а не датсет. Даже комментировать не хочется

##Задание для гениев*


Реализовать расчет корреляции Спирмана

In [ ]:
#Функция по расчету тут

In [ ]:
#Пример использования тут и объяснение принципа работы

###Пример реализации

In [ ]:
import pandas as pd
import numpy as np


def corr_pearson(column_1, column_2):
  cov = np.cov(column_1, column_2)
  return cov/(column_1.std() * column_2.std())

In [ ]:
data = {
    'Рост': [170, 155, 180, 162, 175],
    'Вес': [65, 50, 75, 55, 70]
}
df = pd.DataFrame(data)
df

Рост  Вес
0   170   65
1   155   50
2   180   75
3   162   55
4   175   70

In [ ]:
corr_pearson(df["Рост"], df["Вес"])

array([[0.96593129, 0.99674864],
       [0.99674864, 1.03527033]])